luatest/justrun.lua

```diff
@@ -0,0 +1,164 @@
+local fun = require('fun')
+local json = require('json')
+local fiber = require('fiber')
+
+local log = require('luatest.log')
+
+local has_popen, popen = pcall(require, 'popen')
+if not has_popen then
+    log.error('popen module requires Tarantool 2.4.1 and newer')
+end
+
+local justrun = {}
+
+local function collect_stderr(ph)
+    local f = fiber.new(function()
+        local fiber_name = "child's stderr collector"
+        fiber.name(fiber_name, {truncate = true})
+
+        local chunks = {}
+
+        while true do
+            local chunk, err = ph:read({stderr = true})
+            if chunk == nil then
+                log.warn('%s: got error, exiting: %s', fiber_name, err)
+                break
+            end
+            if chunk == '' then
+                log.info('%s: got EOF, exiting', fiber_name)
+                break
+            end
+            table.insert(chunks, chunk)
+        end
+
+        -- Glue all chunks, strip trailing newline.
+        return table.concat(chunks):rstrip()
+    end)
+    f:set_joinable(true)
+    return f
+end
+
+local function cancel_stderr_fiber(stderr_fiber)
+    if stderr_fiber == nil then
+        return
+    end
+    stderr_fiber:cancel()
+end
+
+local function join_stderr_fiber(stderr_fiber)
+    if stderr_fiber == nil then
+        return
+    end
+    return select(2, assert(stderr_fiber:join()))
+end
+
+-- Run tarantool in given directory with given environment and
+-- command line arguments and catch its output.
+--
+-- Expects JSON lines as the output and parses it into an array
+-- (it can be disabled using `nojson` option).
+--
+-- Options:
+--
+-- - nojson (boolean, default: false)
+--
+--   Don't attempt to decode stdout as a stream of JSON lines,
+--   return as is.
+--
+-- - stderr (boolean, default: false)
+--
+--   Collect stderr and place it into the `stderr` field of the
+--   return value
+--
+-- - quote_args (boolean, default: false)
+--
+--   Quote CLI arguments before concatenating them into a shell
+--   command.
+--
+-- @string dir Directory where the process will run.
+-- @table env Environment variables for the process.
+-- @table args Options that will be passed when the process starts.
+-- @table opts Custom options: nojson, stderr and qoute_args.
+-- @return table
```

Comment:
What is purpose of these luadoc/ldoc comments?

I've generated the HTML pages using `ldoc -d html .` and the module is not present there at all.

If it is just for consistency with other code, then, hm. Don't know. I think that it is weird to write a comment that looks like it adheres ldoc syntax, but it isn't. Moreover, it looks like we generate and deploy the docs to GitHub Pages or some other place. But we don't.

----

I was worrying about `@table` vs `@tparam table`. I'm not sure that the former works as expected, because I had [some workarounds](https://github.com/tarantool/expirationd/blob/419cb99ca60a5c4e6fde570cd39de1c934fdc694/doc/ldoc/config.ld#L11-L13) around in the past.

BTW, it seems you're using `return` with a type, but `treturn` accepts a type.

I don't like `tparam`, `treturn` and so on keywords and prefer the short ones like `string`, `boolean` and so on. But if we're going to use them, we have to add support for them into ldoc's config/template.

Reply:
I agree with you and it's my bad (didn't render the documentation). Now it looks as follows:

![изображение](https://github.com/tarantool/luatest/assets/20374223/a389c1de-6c90-48c9-ab8c-698361968b64)